### DISCUSSION TITLE
Allow zero RPS in create function

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/flow#278] [Ideas]
This is an idea I first shared [here](https://github.com/sablier-labs/flow/discussions/271#discussioncomment-10801555) — not allowing `rps == 0` in the create function is not an objective necessity.

Users may want to set up the streams in advance, and fund them later.

Why not introduce a `adjustRatePerSecondAndDeposit` method to enable this behavior? Although, I think that the `batch` function can also do this, right?

CC @sablier-labs/solidity

**Comment by PaulRBerg** [comment]
To respond to @andreivladbrg's question [here](https://github.com/sablier-labs/flow/discussions/271#discussioncomment-10802766).

> in that case why make 2 separate actions instead of one?

Because of what I said in the OP — users may want to set up the streams in advance so that they can share the URLs with the recipients.

This isn't an incredibly helpful feature, but its existence proves that `rps == 0` is superfluous.

At a minimum, why not remove the check and save a little gas?

**Comment by smol-ninja** [comment]
I am in favor of this proposal. This can also be useful for DAOs where they want to create a stream using Governance, then fund it partially using EOAs and then begin it later (i.e. setting RPS non-zero). There may also be other use cases, but I cant think of them right now.

And why introduce `adjustRatePerSecondAndDeposit` when it can be achieved using the `batch` function? In the [examples repo](https://github.com/sablier-labs/examples) (cc @andreivladbrg), we can include examples to show how users can use `batch` to achieve such composability.

**Comment by andreivladbrg** [comment]
I agree with the proposal. 

re `adjustRatePerSecondAndDeposit`, yes, we can achieve this with `batch`

**Comment by andreivladbrg** [comment]
https://github.com/sablier-labs/flow/issues/279

**Comment by andreivladbrg** [comment]
After seeing @smol-ninja's implementation of this removal, I realized that we actually don't need (as a necessity in the flow contract) the `pause` and `restart` functions, as the same result can be achieved with `adjustRatePerSecond` if we remove the `rps = 0` check.

So, the question that comes to mind is: should we remove them and rely only on `adjustRatePerSecond`? The benefits are:
- Considerably less duplicated code.
- Improved DX, as we have only one function that changes the `rps`.
- Fewer entry points for modifying the state (assuming excessive functions that modify the state can be a security concern).
- Fewer lines of code, leading to a cheaper audit.

wdyt? @sablier-labs/solidity

**Reply by PaulRBerg** [reply]
Disagree.

The DX wouldn't be improve — specific functions with specific roles (even if they're abstractions) are helpful. This is why we have both `withdrawableAmountOf` and `coveredDebtOf`.

> Considerably less duplicated code.

Worthy price to pay

> Fewer entry points for modifying the state

This is what invariant tests are for.

> Fewer lines of code, leading to a cheaper audit.

Negligible

**Reply by andreivladbrg** [reply]
agree, correct

in that case, there is one question left

should we remove `rps = 0` check in `adjustRatePerSecond`?

**Reply by PaulRBerg** [reply]
Yes

**Reply by smol-ninja** [reply]
Happy to remove `rps` check in `adjustRatePerSecond`.

**Reply by andreivladbrg** [reply]
It looks like I forgot to ask about `restart`, sorry:

https://github.com/sablier-labs/flow/blob/6cae74b057b2b7f1b50d31e2bc94b0cd9a3a72db/src/SablierFlow.sol#L706-L714

applying the same logic, we can remove the above checks, correct?

@sablier-labs/solidity

**Reply by smol-ninja** [reply]
If rps can be 0 in restart, then how its even a restart? It would do nothing. Correct? Allowing rps to be 0 in adjustRatePerSecond atleast set it to 0.

**Reply by andreivladbrg** [reply]
> If rps can be 0 in restart, then how its even a restart? It would do nothing. Correct? 

correct, it would do nothing, but why would you restart with an rps = 0? 
removing it saves gas by removing 2 checks, and makes our check consistent across all functions

> Allowing rps to be 0 in adjustRatePerSecond atleast set it to 0.

which is exactly what `pause` function does

**Reply by smol-ninja** [reply]
> which is exactly what the `pause` function does

That's true, but setting rps to 0 via `adjustRatePerSecond` also sets it to zero (doing something), whereas allowing rps to be 0 in restart is doing nothing.

> why would you restart with an rps = 0

You wouldn't but lets say if something does that, it would emit a restart event but wouldn't restart the stream technically. Whereas, `adjustRatePerSecond` with 0 as the new rate per second would adjust it to 0 and emit a relevant event. Its the same argument as if stream is already paused, then `pause` should fail because even a paused stream can be paused again but it would create confusion by doing nothing. What do you think?

**Reply by andreivladbrg** [reply]
>also sets it to zero (doing something), whereas allowing rps to be 0 in restart is doing nothing.

technically it updates the `st` to `block.timestamp`, which means it is doing something

> it would emit a restart event but wouldn't restart the stream technically

the restart event would have `rps = 0` emitted

>Whereas, adjustRatePerSecond with 0 as the new rate per second would adjust it to 0 and emit a relevant event

yes, it would adjust it to zero, which means a pause and we are emitting an `AdjustRatePerSecond` event instead

> then `pause` should fail because even a paused stream can be paused again

good point, that makes me think if we should remove `notPaused` completely

**Reply by andreivladbrg** [reply]
i have an idea, we can remove the `internal` restart function and use `_adjustRatePerSecond` and it would revert due to `SablierFlow_RatePerSecondNotDifferent` if you pass an amount equal to zero (since it is a paused stream which you restart)

**Reply by smol-ninja** [reply]
Sounds good to me.
